File: PRNumberBiz/app/src/main/java/com/pplus/prnumberbiz/apps/product/ProductListFragment.java
//package com.pplus.prnumberbiz.apps.product;
//
//
//import android.content.Context;
//import android.content.Intent;
//import android.graphics.Rect;
//import android.os.Bundle;
//import android.support.annotation.DimenRes;
//import android.support.annotation.NonNull;
//import android.support.v7.widget.GridLayoutManager;
//import android.support.v7.widget.RecyclerView;
//import android.view.View;
//
//import com.pplus.prnumberbiz.Const;
//import com.pplus.prnumberbiz.R;
//import com.pplus.prnumberbiz.apps.common.mgmt.LoginInfoManager;
//import com.pplus.prnumberbiz.apps.common.ui.base.BaseFragment;
//import com.pplus.prnumberbiz.apps.post.ui.PostActivity;
//import com.pplus.prnumberbiz.apps.product.data.ProductAdapter;
//import com.pplus.prnumberbiz.core.code.common.EnumData;
//import com.pplus.prnumberbiz.core.network.ApiBuilder;
//import com.pplus.prnumberbiz.core.network.model.dto.Post;
//import com.pplus.prnumberbiz.core.network.model.response.NewResultResponse;
//
//import java.util.HashMap;
//import java.util.Map;
//
//import network.common.PplusCallback;
//import retrofit2.Call;
//
//public class ProductListFragment extends BaseFragment<PostActivity>{
//
//
//    public ProductListFragment(){
//        // Required empty public constructor
//    }
//
//    public static ProductListFragment newInstance(Long no){
//
//        ProductListFragment fragment = new ProductListFragment();
//        Bundle args = new Bundle();
//        args.putLong(Const.NO, no);
//        fragment.setArguments(args);
//        return fragment;
//    }
//
//    @Override
//    public String getPID(){
//
//        return null;
//    }
//
//    @Override
//    public String getSID(){
//
//        return null;
//    }
//
//    @Override
//    public int getLayoutResourceId(){
//
//        return R.layout.fragment_product_list;
//    }
//
//    private View layout_empty;
//    private ProductAdapter mAdapter;
//    GridLayoutManager mLayoutManager;
//
//    private Long mPageNo;
//    private int mPage;
//    private boolean mLockListView = true, mIsLast = false;
//
//    @Override
//    public void initializeView(View container){
//
//        layout_empty = container.findViewById(R.id.layout_product_config_empty);
//        layout_empty.setVisibility(View.GONE);
//
//        mPageNo = LoginInfoManager.getInstance().getUser().getPage().getNo();
//        RecyclerView recyclerView = (RecyclerView)container.findViewById(R.id.recycler_product_config);
//        mLayoutManager = new GridLayoutManager(getActivity(), 2);
//        recyclerView.setLayoutManager(mLayoutManager);
//        recyclerView.addItemDecoration(new ItemOffsetDecoration(getActivity(), R.dimen.height_150));
//        mAdapter = new ProductAdapter(getActivity());
//        recyclerView.setAdapter(mAdapter);
//
//        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener(){
//
//            int pastVisibleItems, visibleItemCount, totalItemCount;
//
//            @Override
//            public void onScrolled(RecyclerView recyclerView, int dx, int dy){
//
//                super.onScrolled(recyclerView, dx, dy);
//                visibleItemCount = mLayoutManager.getChildCount();
//                totalItemCount = mLayoutManager.getItemCount();
//                pastVisibleItems = mLayoutManager.findFirstVisibleItemPosition();
//                if(!mLockListView) {
//                    if(!mIsLast && (visibleItemCount + pastVisibleItems) >= (totalItemCount * Const.NEXT_RECYCLERVIEW_MARGIN)) {
//                        mPage++;
//                        getPostList(mPage);
//                    }
//                }
//            }
//        });
//
//        mAdapter.setOnItemClickListener(new ProductAdapter.OnItemClickListener(){
//
//            @Override
//            public void onItemClick(int position){
//                Intent intent = new Intent(getActivity(), ProductDetailActivity.class);
//                intent.putExtra(Const.KEY, EnumData.PostType.pr);
//                intent.putExtra(Const.POST_NO, mAdapter.getItem(position).getNo());
//                startActivityForResult(intent, Const.REQ_POST);
//            }
//        });
//
//        mPage = 1;
//        mIsLast = false;
//        getPostList(mPage);
//    }
//
//    @Override
//    public void init(){
//
//    }
//
//    private void getPostList(int page){
//
//        mLockListView = true;
//        Map<String, String> params = new HashMap<>();
//        params.put("no", "" + mPageNo);
//        params.put("filter", EnumData.PostType.pr.name());
//        params.put("pg", "" + page);
//        params.put("sz", "" + Const.LOAD_DATA_LIMIT_CNT);
//
//        showProgress("");
//        ApiBuilder.create().getBoardPostList(params).setCallback(new PplusCallback<NewResultResponse<Post>>(){
//
//            @Override
//            public void onResponse(Call<NewResultResponse<Post>> call, NewResultResponse<Post> getPostList){
//
//                mLockListView = false;
//                hideProgress();
//                if(mPage == 1) {
//                    mAdapter.clear();
//                    if(getPostList.getDatas().size() == 0){
//                        layout_empty.setVisibility(View.VISIBLE);
//                    }else{
//                        layout_empty.setVisibility(View.GONE);
//                    }
//                }
//
//                mAdapter.addAll(getPostList.getDatas());
//                if(getPostList.getDatas() == null || getPostList.getDatas().size() == 0) {
//                    mIsLast = true;
//                    if(mPage > 1){
//                        mPage--;
//                    }
//                }
//            }
//
//            @Override
//            public void onFailure(Call<NewResultResponse<Post>> call, Throwable t, NewResultResponse<Post> response){
//
//                mLockListView = false;
//                hideProgress();
//            }
//        }).build().call();
//    }
//
//    private class ItemOffsetDecoration extends RecyclerView.ItemDecoration{
//
//        private int mItemOffset, mRightOffset;
//
//        public ItemOffsetDecoration(int itemOffset, int rightOffset){
//
//            mItemOffset = itemOffset;
//            mRightOffset = rightOffset;
//        }
//
//        public ItemOffsetDecoration(@NonNull Context context, @DimenRes int itemOffsetId){
//
//            this(context.getResources().getDimensionPixelSize(itemOffsetId), context.getResources().getDimensionPixelSize(R.dimen.width_36));
//        }
//
//        @Override
//        public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state){
//
//            super.getItemOffsets(outRect, view, parent, state);
//
//            int position = parent.getChildAdapterPosition(view);
//
//            if(position % 2 == 1){
//                outRect.set(0, 0, 0, mItemOffset);
//            }else{
//                outRect.set(0, 0, mRightOffset, mItemOffset);
//            }
//
//        }
//    }
//}

File: Buflexz/app/src/main/java/com/root37/buflexz/apps/product/ui/ProductListFragment.kt
package com.root37.buflexz.apps.product.ui

import android.content.Context
import android.content.Intent
import android.graphics.Rect
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.DimenRes
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.pplus.networks.common.PplusCallback
import com.root37.buflexz.Const
import com.root37.buflexz.R
import com.root37.buflexz.apps.common.ui.base.BaseActivity
import com.root37.buflexz.apps.common.ui.base.BaseFragment
import com.root37.buflexz.apps.main.ui.MainProductFragment
import com.root37.buflexz.apps.product.data.ProductAdapter
import com.root37.buflexz.core.network.ApiBuilder
import com.root37.buflexz.core.network.model.dto.Product
import com.root37.buflexz.core.network.model.dto.ProductCategory
import com.root37.buflexz.core.network.model.response.ListResultResponse
import com.root37.buflexz.core.network.model.response.NewResultResponse
import com.root37.buflexz.core.util.PplusCommonUtil
import com.root37.buflexz.databinding.FragmentProductListBinding
import retrofit2.Call

class ProductListFragment : BaseFragment<BaseActivity>() {
    private var mCategory: ProductCategory? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            mCategory = PplusCommonUtil.getParcelable(it, Const.CATEGORY, ProductCategory::class.java)
        }
    }

    private var _binding: FragmentProductListBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        _binding = FragmentProductListBinding.inflate(inflater, container, false)
        val view = binding.root
        return view
    }

    override fun getPID(): String? {
        return ""
    }

    private var mTotalCount = 0
    var mAdapter: ProductAdapter? = null
    private lateinit var mLayoutManager: GridLayoutManager
    private var mLockListView = false
    private var mPaging = 1
    private var mDir = "ASC"

    override fun init() {
        mLayoutManager = GridLayoutManager(requireActivity(), 2)
        binding.recyclerProductList.layoutManager = mLayoutManager
        mAdapter = ProductAdapter()
        binding.recyclerProductList.adapter = mAdapter
        binding.recyclerProductList.addItemDecoration(CustomItemOffsetDecoration(requireActivity(), R.dimen.width_33, R.dimen.height_69))

        binding.recyclerProductList.addOnScrollListener(object : RecyclerView.OnScrollListener() {

            internal var pastVisibleItems: Int = 0
            internal var visibleItemCount: Int = 0
            internal var totalItemCount: Int = 0

            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {

                super.onScrolled(recyclerView, dx, dy)

                visibleItemCount = mLayoutManager.childCount
                totalItemCount = mLayoutManager.itemCount
                pastVisibleItems = mLayoutManager.findFirstVisibleItemPosition()
                if (!mLockListView) {
                    if (totalItemCount < mTotalCount && visibleItemCount + pastVisibleItems >= totalItemCount * Const.NEXT_RECYCLERVIEW_MARGIN) {
                        mPaging++
                        getList(mPaging)
                    }
                }
            }
        })

        mAdapter!!.listener = object : ProductAdapter.OnItemClickListener {
            override fun onItemClick(position: Int) {
                val intent = Intent(requireActivity(), ProductActivity::class.java)
                intent.putExtra(Const.DATA, mAdapter!!.getItem(position))
                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                defaultLauncher.launch(intent)
            }
        }

        mPaging = 1
        getList(mPaging)
    }

    private fun getList(page: Int) {
        val params = HashMap<String, String>()
        params["paging[page]"] = page.toString()
        params["paging[limit]"] = "20"
        if (mCategory!!.seqNo != null) {
            params["categorySeqNo"] = mCategory!!.seqNo.toString()
        } //        params["order[][column]"] = "price"
        //        params["order[][dir]"] = mDir
        showProgress("")
        mLockListView = true
        ApiBuilder.create().getProductList(params).setCallback(object : PplusCallback<NewResultResponse<ListResultResponse<Product>>> {
            override fun onResponse(call: Call<NewResultResponse<ListResultResponse<Product>>>?, response: NewResultResponse<ListResultResponse<Product>>?) {
                hideProgress()
                if (response?.result != null) {
                    if (page == 1) {
                        mAdapter!!.clear()

                        mTotalCount = response.result!!.total!!
                        if(mTotalCount == 0){
                            binding.textProductListNotExist.visibility = View.VISIBLE
                        }else{
                            binding.textProductListNotExist.visibility = View.GONE
                        }
                    }

                    mLockListView = false

                    if (response.result!!.list != null) {
                        val dataList = response.result!!.list!!
                        mAdapter!!.addAll(dataList)
                    }

                }
            }

            override fun onFailure(call: Call<NewResultResponse<ListResultResponse<Product>>>?, t: Throwable?, response: NewResultResponse<ListResultResponse<Product>>?) {
                hideProgress()
            }
        }).build().call()
    }

    private val defaultLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        mPaging = 1
        getList(mPaging)
        if (getParentActivity() is ProductCategoryActivity) {
            (getParentActivity() as ProductCategoryActivity).loginCheck()
        } else if (parentFragment is MainProductFragment) {
            (parentFragment as MainProductFragment).loginCheck()
        }
    }

    private inner class CustomItemOffsetDecoration(private val mRightOffset: Int, private val mBottomOffset: Int) : RecyclerView.ItemDecoration() {

        constructor(context: Context, @DimenRes rightOffsetId: Int, @DimenRes bottomOffsetId: Int) : this(context.resources.getDimensionPixelSize(rightOffsetId), context.resources.getDimensionPixelSize(bottomOffsetId)) {
        }

        override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {

            super.getItemOffsets(outRect, view, parent, state)

            val position = parent.getChildAdapterPosition(view)

            if (position % 2 != 1) {
                outRect.set(0, 0, mRightOffset, mBottomOffset)
            } else {
                outRect.set(0, 0, 0, mBottomOffset)
            }
        }
    }

    companion object {

        @JvmStatic
        fun newInstance(category: ProductCategory) = ProductListFragment().apply {
            arguments = Bundle().apply {
                putParcelable(Const.CATEGORY, category)
            }
        }
    }
}